### ISSUE TITLE
What are some errors you had in the project (especially level 3)

### ISSUE THREAD

**Issue by Hongyi6328** [nus-cs2030/2122-s1#188]
# Summary
I am currently at the midst of doing the project but I encountered a test case that I cannot pass for a reason that I have no idea about. I would like to to hear some experiences and reflect on my project.

## Description 
Describe the problem and code snippets if necessary 

## Screenshots (if any): 
Insert Images here if necessary

**Comment by RussellDash332** [comment]
> # Summary
> I am currently in the midst of doing the project but I encountered a test case that I cannot pass for a reason that I have no idea about. I would like to to hear some experiences and reflect on my project.
> 

What kind of error did you experience?

**Comment by thojianxiong** [comment]
I think I was stuck on Level 3 for quite some time because my thought process for the flow of events was not very clear? I was confused between the difference in executing the actual service of the customer and when does the SERVE event occur, due to the addition of server rest. In the end, I managed to visualise the customer flow and only added the SERVE event when the server is actually free to serve after resting.

**Comment by RussellDash332** [comment]
Anyways, if you need to see how the test case is visualized, head over to #115

**Comment by JordanChua** [comment]
Just to add on since I was also stuck on Level 3 and had to change how my implementation worked, I think the main issue here is probably understanding when the Serve Event is executed. Previously I made the mistake of assuming that once a customer arrives we will know exactly when he will be served, but thats not necessarily true since the server can now rest, and we now need to account for the rest time of the server which we can only know when the server is actually done serving a customer and is able to rest. Ás mentioned above we should ONLY serve the customer once the server is done resting and is available, and we shouldn't be assuming when the server will start serving the customer from the point of arrival.

I originally only had 1 Rest Event, but I had the problem where I couldn't update the status of server once hes actually done resting. One way to get about this is to simply introduce a new RestDone Event to update the server (so now it is available to serve), but there are definitely other possible solutions as well :)

**Comment by chenghong99** [comment]
The way i handled level 3 was to create 2 addition events. ServerResting and ServerFinishResting. One is for the server to start rest depending on the rest time. The other is for the server to go back to serve. Since all levels are already out you should know by now that everything has to be lazily evaluated.

**Comment by weiquany** [comment]
Instead of creating new events, a possible way is to push Wait events back when the customer is not next to be served and only create the Serve event when the server is not resting and is available to serve the customer next.

**Comment by FYimu** [comment]
I made two very stupid mistakes that took me quite a long time to debug. The first one was for Servers that do not need to rest, I created a Rest event of 0 restTime for them and rest event has lower priority than arrive event. And the second one was I forgot to overwrite a method for SelfCheckout class.

One way of debugging that I found very effective (tho painful and make my eyes sore) is to print out the status of all servers' currently serving customer, next customer in the queue, and number of customers in the queue. It helps you to observe which part of the code (which class and which method) might have errors. 
This was how I found the wait() of my SelfCheckout magically eats customer in the queue XP

**Comment by Hongyi6328** [comment]
OK finally I've got some experience to share after I realized my own errors.
Though the project instructions specifies that all SelfCheckOuts share the same queue, but I think it'd be good for a SelfCheckOut to have an individual queue. I eventually used a class called SelfCheckOutGroup to handle the wait, serve, done, etc requests between Customer and SelfCheckOut. 
The time does not flow continuously in computer, every time stamp is discrete, so I don't think there is a need to create a global variable that ticks as time. But instead, I think every server has to have its own "reference time stamp" that specifies when exactly it starts to server the currently serving customer, so you can easily determine when is its "next available time" to server another customer.
Level 5 seemed to have frightened me because I thought the way I implemented the first 4 levels was not compatible with lazy evaluation. I thought to be lazy means a total rework of the main architecture of my project to make it something like the mini-project, but after some thinking I realised that being lazy does not necessarily mean you need suppliers. You can actually postpone the determination of WHEN to serve to the moment that it can be decided. So when a customer arrives, the only decision you make is whether to let it leave or stay. This can be done eagerly because you know the length of the queues without knowing the service time of this customer.
